File: apps/plant_shop_c-sharp/plant_shop_asp_dapper/Repositories/PlantRepository.cs
using Dapper;
using plant_shop_asp_dapper.Data;
using plant_shop_asp_dapper.Models;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace plant_shop_asp_dapper.Repositories
{
    /// <summary>
    /// Repository CRUD pour les plantes.
    /// </summary>
    public class PlantRepository : BaseRepository
    {
        /// <summary>
        /// Constructeur avec injection de la factory.
        /// </summary>
        /// <param name="factory">Factory de connexions.</param>
        public PlantRepository(DbConnectionFactory factory) : base(factory) { }

        private const string SelectSql = @"
            SELECT id AS Id,
                   name AS Name,
                   description AS Description,
                   price AS Price,
                   stock AS Stock,
                   created_at AS CreatedAt
            FROM plants";

        /// <summary>
        /// Trouve une plante par ID.
        /// </summary>
        /// <param name="id">ID a rechercher.</param>
        /// <returns>Plante ou null.</returns>
        public async Task<Plant?> FindByIdAsync(int id)
        {
            using var connection = CreateConnection();
            return await connection.QuerySingleOrDefaultAsync<Plant>(
                $"{SelectSql} WHERE id = @Id", new { Id = id });
        }

        public async Task<IEnumerable<Plant>> FindAllAsync(bool includeOutOfStock = false)
        {
            using var connection = CreateConnection();
            var sql = includeOutOfStock
                ? $"{SelectSql} ORDER BY name ASC"
                : $"{SelectSql} WHERE stock > 0 ORDER BY name ASC";

            return await connection.QueryAsync<Plant>(sql);
        }

        /// <summary>
        /// Cree une nouvelle plante.
        /// </summary>
        /// <param name="plant">Plante a creer.</param>
        /// <returns>Plante avec ID genere.</returns>
        public async Task<Plant> CreateAsync(Plant plant)
        {
            using var connection = CreateConnection();
            var sql = @"
                INSERT INTO plants (name, description, price, stock, created_at)
                VALUES (@Name, @Description, @Price, @Stock, @CreatedAt)
                RETURNING id";

            plant.CreatedAt = DateTime.UtcNow;
            plant.Id = await connection.ExecuteScalarAsync<int>(sql, plant);
            return plant;
        }

        /// <summary>
        /// Met a jour une plante.
        /// </summary>
        /// <param name="plant">Plante modifiee.</param>
        public async Task UpdateAsync(Plant plant)
        {
            using var connection = CreateConnection();
            var sql = @"
                UPDATE plants
                SET name = @Name,
                    description = @Description,
                    price = @Price,
                    stock = @Stock
                WHERE id = @Id";
            await connection.ExecuteAsync(sql, plant);
        }

        public async Task<int> UpdateStockAsync(int id, int newStock)
        {
             using var connection = CreateConnection();
             return await connection.ExecuteAsync(
                "UPDATE plants SET stock = @Stock WHERE id = @Id",
                new { Id = id, Stock = newStock });
        }


        /// <summary>
        /// Supprime une plante.
        /// </summary>
        /// <param name="id">ID a supprimer.</param>
        public async Task DeleteAsync(int id)
        {
            using var connection = CreateConnection();
            await connection.ExecuteAsync("DELETE FROM plants WHERE id = @Id", new { Id = id });
        }
    }
}

File: apps/plant_shop_java/Structure_en_micro_services/plant_shop_java_spring_microservices/plant_shop_java_spring_boot_security_microservices/catalog-service/src/repository/PlantRepository.java
package repository;

import model.Plant;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Repository;
import org.springframework.web.context.annotation.RequestScope;
import java.sql.*;
import java.util.ArrayList;
import java.util.List;

/**
 * Repository pour la gestion des plantes en base de données.
 * Fournit les opérations CRUD sur la table plants.
 */
@Repository
@RequestScope
public class PlantRepository {

    private final Connection db;

    /**
     * Constructeur avec injection de la connexion BDD.
     * @param db Connexion à la base de données
     */
    @Autowired
    public PlantRepository(Connection db) {
        this.db = db;
    }

    /**
     * Recherche une plante par son identifiant.
     * @param id Identifiant de la plante
     * @return La plante trouvée ou null
     * @throws SQLException En cas d'erreur SQL
     */
    public Plant find(int id) throws SQLException {
        String sql = "SELECT * FROM plants WHERE id=?";
        try (PreparedStatement ps = db.prepareStatement(sql)) {
            ps.setInt(1, id);
            try (ResultSet rs = ps.executeQuery()) {
                if (rs.next()) {
                    return mapFromResultSet(rs);
                }
                return null;
            }
        }
    }

    /**
     * Récupère toutes les plantes du catalogue.
     * @return Liste des plantes triées par id
     * @throws SQLException En cas d'erreur SQL
     */
    public List<Plant> findAll() throws SQLException {
        String sql = "SELECT * FROM plants ORDER BY id";
        List<Plant> plants = new ArrayList<>();
        try (PreparedStatement ps = db.prepareStatement(sql);
             ResultSet rs = ps.executeQuery()) {
            while (rs.next()) {
                plants.add(mapFromResultSet(rs));
            }
        }
        return plants;
    }

    /**
     * Supprime une plante par son identifiant.
     * @param id Identifiant de la plante à supprimer
     * @throws SQLException En cas d'erreur SQL
     */
    public void delete(int id) throws SQLException {
        String sql = "DELETE FROM plants WHERE id=?";
        try (PreparedStatement ps = db.prepareStatement(sql)) {
            ps.setInt(1, id);
            ps.executeUpdate();
        }
    }

    /**
     * Mappe un ResultSet vers un objet Plant.
     * @param rs ResultSet positionné sur une ligne
     * @return La plante mappée
     * @throws SQLException En cas d'erreur SQL
     */
    private Plant mapFromResultSet(ResultSet rs) throws SQLException {
        return new Plant(
            rs.getInt("id"),
            rs.getString("name"),
            rs.getString("description"),
            rs.getBigDecimal("price"),
            rs.getInt("stock"),
            rs.getTimestamp("created_at")
        );
    }

    /**
     * Crée une nouvelle plante en base de données.
     * @param p Plante à créer
     * @return L'identifiant généré
     * @throws SQLException En cas d'erreur SQL
     */
    public int create(Plant p) throws SQLException {
        String sql = "INSERT INTO plants(name, description, price, stock) VALUES (?, ?, ?, ?)";
        try (PreparedStatement ps = db.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS)) {
            ps.setString(1, p.name);
            ps.setString(2, p.description != null ? p.description : ""); // Gère null
            ps.setBigDecimal(3, p.price);
            ps.setInt(4, p.stock);
            ps.executeUpdate();
            try (ResultSet rs = ps.getGeneratedKeys()) {
                rs.next();
                return rs.getInt(1);
            }
        }
    }

    /**
     * Met à jour une plante existante.
     * @param p Plante avec les nouvelles valeurs
     * @throws SQLException En cas d'erreur SQL
     */
    public void update(Plant p) throws SQLException {
        String sql = "UPDATE plants SET name=?, description=?, price=?, stock=? WHERE id=?";
        try (PreparedStatement ps = db.prepareStatement(sql)) {
            ps.setString(1, p.name);
            ps.setString(2, p.description);
            ps.setBigDecimal(3, p.price);
            ps.setInt(4, p.stock);
            ps.setInt(5, p.id);
            ps.executeUpdate();
        }
    }

    /**
     * Met à jour le stock d'une plante.
     * @param id Identifiant de la plante
     * @param newStock Nouvelle quantité en stock
     * @throws SQLException En cas d'erreur SQL
     */
    public void updateStock(int id, int newStock) throws SQLException {
        String sql = "UPDATE plants SET stock=? WHERE id=?";
        try (PreparedStatement ps = db.prepareStatement(sql)) {
            ps.setInt(1, newStock);
            ps.setInt(2, id);
            ps.executeUpdate();
        }
    }
}
